python-spec/src/somacore/coordinates.py

```diff
@@ -0,0 +1,445 @@
+
+        # Create and validate the augmented matrix.
+        self._matrix: npt.NDArray[np.float64] = np.array(matrix, dtype=np.float64)
+        if self._matrix.shape == (rank + 1, rank + 1):
+            if not (
+                self._matrix[-1, -1] == 1.0
+                and np.array_equal(self._matrix[-1, :-1], np.zeros((rank,)))
+            ):
+                raise ValueError(
+                    f"Input matrix {self._matrix} has augmented matrix shape, but is not a valid "
+                    f"augmented matrix."
+                )
+        elif self._matrix.shape == (rank, rank + 1):
+            self._matrix = np.vstack(
+                (
+                    self._matrix,
+                    np.hstack((np.zeros((rank,)), np.array([1]))),
+                )
+            )
+        elif self._matrix.shape == (rank, rank):
+            self._matrix = np.vstack(
+                (
+                    np.hstack((self._matrix, np.zeros((rank, 1)))),
+                    np.hstack((np.zeros((rank,)), np.array([1]))),
+                )
+            )
+        else:
+            raise ValueError(
+                f"Unexpected shape {self._matrix.shape} for the input affine matrix."
+            )
+
+    def __matmul__(self, other: object) -> CoordinateTransform:
+        if not isinstance(other, CoordinateTransform):
+            raise NotImplementedError(
+                f"Matrix multiply is not implemented with type {type(other)!r}."
+            )
+        self._check_matmul_inner_axes(other)
+        if self.input_axes != other.output_axes:
+            raise ValueError("Axis mismatch between transformations.")
+        if isinstance(other, IdentityTransform):
+            return AffineTransform(other.input_axes, self.output_axes, self._matrix)
+        if isinstance(other, AffineTransform):
+            return AffineTransform(
+                other.input_axes,
+                self.output_axes,
+                self.augmented_matrix @ other.augmented_matrix,
+            )
+        raise NotImplementedError(
+            f"Cannot multiply a CoordinateTransform by type {type(other)!r}."
+        )
+
+    def __rmatmul__(self, other: object) -> CoordinateTransform:
+        if not isinstance(other, CoordinateTransform):
+            raise NotImplementedError(
+                f"Matrix multiply is not implemented with type {type(other)!r}."
+            )
+        self._check_rmatmul_inner_axes(other)
+        if isinstance(other, IdentityTransform):
+            return AffineTransform(
+                self.input_axes, other.output_axes, self.augmented_matrix
+            )
+        if isinstance(other, AffineTransform):
+            return AffineTransform(
+                self.input_axes,
+                other.output_axes,
+                other.augmented_matrix @ self.augmented_matrix,
+            )
+        raise NotImplementedError(
+            f"Cannot multiply a CoordinateTransform by type {type(other)!r}."
+        )
```

Comment:
looking again at the logic here: right now there is a certain amount of duplicated logic between the `__matmul__` and the `__rmatmul__` functions. Would it make sense to formulate the matrix multiplications _entirely_ based on `__matmul__`? That would make the code that is executed a bit more predictable, so that `a @ b` will _always_ be `a.__matmul__(b)`, and `a` only has to worry about all the possible `b`s, but `b` doesn’t need to know anything special about `a`. In most circumstances, `b.__rmatmul__(a)` will never be called anyway, because within our little ecosystem, all `a`s know about all `b`s.